Make it possible to build GuideML on Haiku

pulkomandy opened the pull request:
Hello,

This is my fork of GuideML to make it usable on Haiku. It uses a set of wrapper headers that allow me to re-route the Amiga standard libraries to the Haiku equivalents.

**Do not merge this in its current state**: it currently breaks the build for Amiga systems. My goal in opening this pull request is to know if you are interested in merging this in a cleaned form, in which case I will spend time cleaning this up to avoid breaking the build on Amiga by using proper ifdefs and other needed adjustments. Otherwise, I can keep rebasing and adjsuting my port when you update the Amiga version.

pulkomandy committed:
Initial work on porting to Haiku

pulkomandy committed:
Parse 'system "openurl ..."' and resolve it with just the URL.
Not very solid parsing, but that should be ok for my use...

pulkomandy committed:
Tweak the colors a little

pulkomandy committed:
Fix 64-bit compatibility

pulkomandy committed:
Merge guideml 3.16 with Haiku work

pulkomandy committed:
Fix Haiku build

- Add several empty include files to wrapper library
- ifdef out all the GUI code when building for Haiku, as well as the
  locale support

chris-y commented:
I'm happy to merge this in if it is cleaned up and still builds for OS4.

pulkomandy committed:
Cleaup, try to fix AmigaOS build

pulkomandy commented on guideml.c:
```diff
@@ -908,127 +922,164 @@ if((param.smartwrap) && (strlen(buf)<2))
           link++;
           while(*link && (*link==' ' || *link==',' || *link=='\t')) link++;
           if(!*link) return(0);
-          if(Strnicmp(link,"link",4))
```
Functionnal change: I have added some handling of the "system" command when it is used with the "openurl". This is used in the ace.cpcscene.net userguide for links to external websites. This code converts them to html links to the corresponding URLs.

The code is a bit incomplete, in particular it does not check if "system" is used with something else than the "openurl" command and will probably generate invalid links in that case.

chris-y replied:
Some checks to see if it is using openurl would be useful here, although I'm not sure the system command sees much use.

pulkomandy replied:
I added a check for openurl, in other cases the command is ignored with a warning message and no link is generated.

pulkomandy commented on guideml.c:
```diff
@@ -2678,15 +2741,15 @@ if(!wb)
-	param.colours[4] = 0x000000;
-	param.colours[5] = 0xcfcfcf;
-	param.colours[6] = 0xffffff;
+	param.colours[0] = 0x000000; // text
```
Functional change: the default colors are now black on white. The "shine" color becomes red/pink instead of white and the "highlight" color becomes light blue instead of white.

chris-y replied:
The original colours were picked to mimic AmigaGuide. I'm not too fussed about this. A defaults toggle in the GUI might be preferable? At one point I was considering a dark mode option which would also work here. Original/Light/Dark buttons to set the defaults maybe?

pulkomandy replied:
I see that the colors are set twice, and I changed only the command-line version. So if you use the GUI you would still get the AmigaGuide colors.

Also when using the GUI, these can already be changed using the COLOUR0 to COLOUR6 tooltypes, but there are no matching command-line parameters (which is why I ended up hardcoding my preferred colors).

I will not try to update the GUI myself (no way to test it), I could try to add the COLOUR0 to COLOUR6 parameters in readargs, maybe.

chris-y replied:
I didn't realise they were set twice - this wasn't originally my program and I can't remember what I did when I added the GUI (twenty years or so ago...). Some command line overrides would be fab. I will do some presets in the GUI once it's all merged.

pulkomandy replied:
Ok, there is now a THEME option with a choice between AmigaGuide, Light and Dark themes. And the color list is in a single place.

chris-y replied:
Fab, thanks.

pulkomandy reviewed (commented):
I think this version should mostly work, but I probably missed some things. Apparently you need to allow the continuous integration to run and see if it's all OK (I don't have an Amiga system available)

I expect some problems with the use of IPTR (not available in AmigaOS 3, it should be defined to LONG there) and UIPTR (I just made this one up, it should be defined to ULONG on 32bit systems and uintptr_t for me on Haiku to allow 64-bit support). The defines for AmigaOS should be added if needed.

All other changes are now just disabling some parts of the code for Haiku with ifdefs. Except two functional changes which I have highlighted with inline comments, let me know what you think about these.

chris-y commented:
I've allowed the CI to run, can you see the log?

pulkomandy committed:
Verify that system commands do use "openurl"

Other system commands are skipped (the text is copied without any link)
with a warning message.

Also another attempt to fix the AmigaOS build by defining IPTR and UIPTR
since they don't exist there (they seem to be an AROS addition for 64bit
safety)

pulkomandy committed:
Add simple colortheme system

The theme can be switched using the command line THEME parameter. Not
accessible from the GUI yet.

chris-y commented:
Can you bump the version number and add some comments to the changelog too?

pulkomandy committed:
Change version and release date

chris-y commented:
Build errors:
OS4:
guideml.c: In function 'wbmain':
guideml.c:3039: error: conflicting types for 'i'
guideml.c:3020: error: previous declaration of 'i' was here


OS3:
> CloseLibrary(GetColorBase);
error 82 in line 5196 of "guideml.c": unknown identifier <GetColorBase>

I suspect the OS3 one isn't due to anything you've done.

pulkomandy committed:
More AmigaOS fixes

pulkomandy commented:
They should both be fixed now

chris-y commented:
Merged, thanks.

I'll add some theme presets to the GUI and do a new release.
